crypto/hash/xor_unaligned_asbytes_norace.go

```diff
@@ -0,0 +1,47 @@
+/// The functions below were copied and modified from golang.org/x/crypto/sha3.
+//
+// Copyright (c) 2009 The Go Authors. All rights reserved.
+
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are
+// met:
+
+//    * Redistributions of source code must retain the above copyright
+// notice, this list of conditions and the following disclaimer.
+//    * Redistributions in binary form must reproduce the above
+// copyright notice, this list of conditions and the following disclaimer
+// in the documentation and/or other materials provided with the
+// distribution.
+//    * Neither the name of Google Inc. nor the names of its
+// contributors may be used to endorse or promote products derived from
+// this software without specific prior written permission.
+
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+// "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+// LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+// A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+// OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+// LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+// DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+// THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+//go:build (amd64 || 386 || ppc64le) && !purego && !race
+// +build amd64 386 ppc64le
+// +build !purego
+// +build !race
+
+package hash
+
+import "unsafe"
+
+func (b *storageBuf) asBytes() *[maxRate]byte {
+	// re-using a trick from https://github.com/golang/go/blob/master/src/runtime/stubs.go#L178:
+	// to hide the input pointer from escape analysis and avoid
+	// an escape to to the heap. The 0 xor tricks the escape analysis tool
+	// to think "ptr" and "b" are not related.
+	ptr := uintptr(unsafe.Pointer(b)) ^ 0 // nolint:staticcheck
```

Comment:
I'm curious what kind of race conditions this line has caused, and if there are ways to avoid them other than removing the line?

Reply:
[race detector implicitly](https://go-review.googlesource.com/c/go/+/201783/) enables `checkptr` so [it fails with the following](https://go.dev/play/p/vc0QHhoOZrY?v=gotip):
```
$ go run -race main.go
fatal error: checkptr: pointer arithmetic result points to invalid allocation

goroutine 1 [running]:
runtime.throw({0x4edf95?, 0x4906d9?})
	/usr/lib/go/src/runtime/panic.go:992 +0x71 fp=0xc000119ec8 sp=0xc000119e98 pc=0x462951
runtime.checkptrArithmetic(0x0?, {0x0, 0x0, 0xc000119f70?})
	/usr/lib/go/src/runtime/checkptr.go:69 +0xaa fp=0xc000119ef8 sp=0xc000119ec8 pc=0x437d2a
main.(*storageBuf).asBytes(...)
```

It is possible to disable it with `-d=checkptr=0` but given that race detector enabled build is already quite slow, I considered that it makes sense to just disable this optimization there.

Reply:
(Alternatively, i can add `//go:nocheckptr` to the function, which may be more obvious)

Reply:
Yeah, you are totally right: adding an annotation is way more straightforward:  https://github.com/onflow/flow-go/pull/2293/commits/1288c1dbcfa19c4249cdc0ae88078e7f4112b4d5

Reply:
I like the `//go:nocheckptr` idea. I tested and it works well 👌🏼. Thank you for the details!